java/src/hu/holyoil/view/panels/EnvironmentPanel.java

```diff
@@ -60,43 +126,104 @@ else if (res.IsSameType(new Water()))
         return image;
     }
 
+    /**
+     * Segít megállapítani, melyik képre van szüksége az játékban lévő entitások képei közül.
+     * @param spaceship entitás
+     * @return entitás képe
+     */
+    private Image DefineImageFrom(AbstractSpaceship spaceship) {
+        return spaceship.GetImage();
+    }
+
+    /**
+     * Képernyő invalidációja során lefutó metódus, amely override-ja szükséges a grafikus elemek kirajzolására.
+     * @param g grafikus objektum
+     */
     @Override
     protected void paintComponent(Graphics g) {
         super.paintComponent(g);
 
+        // don't query player before initialized!
         if (!canDraw)
             return;
 
+        // draw all asteroids neighbouring
         List<Asteroid> asteroids = player.GetOnAsteroid().GetNeighbours();
-        for (int i = 0; i < asteroids.size(); ++i) {
-            double phi = 2 * Math.PI * i / asteroids.size();
-            int x = (int) (Math.cos(phi) * 280) + 330;
-            int y = (int) (Math.sin(phi) * 220) + 260;
-            g.drawImage(asteroidImg, x, y, 42, 42, this);
-
-            AbstractBaseResource res = asteroids.get(i).GetResource();
-            if (res == null)
-                continue;
-            g.drawImage(DefineImageFrom(res), x + 1, y + 1, 40, 40, this);
+        for (Asteroid asteroid : asteroids) {
+            Point point = asteroidPointMap.get(asteroid);
+            g.drawImage(asteroidImg, point.x, point.y, 44, 44, this);
+
+            AbstractBaseResource res = asteroid.GetResource();
+            if (res != null)
+                g.drawImage(DefineImageFrom(res), point.x + 2, point.y + 2, 40, 40, this);
+
+            // draw spaceships
+            List<AbstractSpaceship> spaceships = asteroid.GetSpaceships();
+
+            double phi = -0.5 * Math.PI;
+            double deltaPhi = 2 * Math.PI / (double)spaceships.size();
+            int x = (int) (Math.cos(phi) * 32) + point.x + 12;
+            int y = (int) (Math.sin(phi) * 32) + point.y + 12;
+
+            if (asteroid.GetTeleporter() != null) {
+                g.drawImage(teleportImg, x, y, 20, 20, this);
+                deltaPhi = 2 * Math.PI / (double)(spaceships.size() + 1);
+                phi += deltaPhi;
+            }
+
+            for (AbstractSpaceship spaceship : spaceships) {
+                x = (int) (Math.cos(phi) * 32) + point.x + 12;
+                y = (int) (Math.sin(phi) * 32) + point.y + 12;
+                g.drawImage(DefineImageFrom(spaceship), x, y, 20, 20, this);
+                phi += deltaPhi;
+            }
```

Comment:
Drawing out an asteroid (central or neighbour alike) could be moved into a separate function. This would clean up the code nicely

Reply:
Done